flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncCollectorBuffer.java

```diff
@@ -0,0 +1,494 @@
+
+					notFull.signal();
+				}
+				catch (Exception e) {
+					error = e;
+					break;
+				}
+			}
+		}
+
+		/**
+		 * Emit results for each finished collector.
+		 */
+		private void unorderedProcess() {
+			AsyncCollector collector = finishedCollectors.pollFirst();
+			while (collector != null) {
+				try {
+					output(collector);
+
+					queue.remove(collectorToQueue.get(collector));
+					clearInfoInMaps(collector);
+
+					notFull.signal();
+
+					collector = finishedCollectors.pollFirst();
+				}
+				catch (Exception e) {
+					error = e;
+					break;
+				}
+			}
+		}
+
+		/**
+		 * If some bad things happened(like exceptions from async i/o), the operator tries to fail
+		 * itself at:
+		 *   {@link AsyncWaitOperator#processElement}, triggered by calling {@link AsyncCollectorBuffer#add}.
+		 *   {@link AsyncWaitOperator#snapshotState}
+		 *   {@link AsyncWaitOperator#close} while calling {@link AsyncCollectorBuffer#waitEmpty}
+		 *
+		 * It is necessary for Emitter Thread to notify methods blocking on notFull/isEmpty.
+		 */
+		private void processError() {
+			queue.clear();
+			finishedCollectors.clear();
+			collectorToQueue.clear();
+			collectorToStreamElement.clear();
+
+			notFull.signalAll();
+			isEmpty.signalAll();
+		}
+
+		/**
+		 * If
+		 *   In ordered mode, there are some finished async collectors, and one of them is the first element in
+		 *   the queue.
+		 * or
+		 *   In unordered mode, there are some finished async collectors.
+		 * or
+		 *   The first element in the queue is Watermark.
+		 * Then, the emitter thread should keep waiting, rather than waiting on the condition.
+		 *
+		 * Otherwise, the thread should stop for a while until being signalled.
+		 */
+		private boolean nothingToDo() {
+			if (queue.size() == 0) {
+				isEmpty.signalAll();
+				return true;
+			}
+
+			// while doing checkpoints, emitter thread should not try to output records.
+			if (isCheckpointing) {
+				return true;
+			}
+
+			// check head element of the queue, it is OK to process Watermark or LatencyMarker
+			if (collectorToStreamElement.get(queue.get(0)).isWatermark() ||
+				collectorToStreamElement.get(queue.get(0)).isLatencyMarker())
+			{
+				return false;
+			}
+
+			if (mode == AsyncDataStream.OutputMode.UNORDERED) {
+				// no finished async collector...
+				if (finishedCollectors.size() == 0) {
+					return true;
+				}
+				else {
+					return false;
+				}
+			}
+			else {
+				// for ORDERED mode, make sure the first collector in the queue has been done.
+				AsyncCollector collector = queue.get(0);
+				if (collector.isDone() == false) {
+					return true;
+				}
+				else {
+					return false;
+				}
+			}
+		}
+
+		@Override
+		public void run() {
+			while (running) {
+				try {
+					lock.lock();
+
+					if (error != null) {
+						// stop processing finished async collectors, and try to wake up blocked main
+						// thread or checkpoint thread repeatedly.
+						processError();
```

Comment:
It has to signal main thread which may wait there while trying to close async wait operator. When task tries to close all operators, it has to wait for while until all `AsyncCollector` have finished. Assume that some error happens first, `processError()` is called and Emitter thread exits, then main thread finds the buffer is not empty, it will wait there permanently.